### PR TITLE
[Translation][Portuguese Brazil] arrays.article

### PR DESCRIPTION
**Pull request by sergiolpf**

### CODE REVIEW

File: _content/tour/por/arrays.article [hunk 1]
```diff
@@ -1,163 +1,153 @@
...
 
-When using pointer semantic iteration, you iterate over the original collection and I
-access each element associated with the collection directly.
+Quando iteramos usando ponteiros, você itera sobre a collection original e acessa diretamente cada elemento associado com a collection. 
 
-** Value Semantic Iteration
+** Iteração por valor
 
-Given the following code and output.
+Dado o seguinte code e seu resultado
```
Comment (rschio) [inline _content/tour/por/arrays.article:74]:
```suggestion
Dado o seguinte código e seu resultado
```

File: _content/tour/por/arrays.article [hunk 2]
```diff
@@ -1,163 +1,153 @@
...
 
-When using value semantic iteration, two things happen. First, the collection I’m
-iterating over is copied and you iterate over the copy. In the case of an array, the
-copy could be expensive since the entire array is copied. In the case of a slice,
-there is no real cost since only the internal slice value is copied and not the
-backing array. Second, you get a copy of each element being iterated on.
+Quando iteramos usando valor, duas coisas acontecem. Primeira, a collection inteira é copiada e você navega sobre a copia.
+No caso de arrays, a copia pode ser bem cara uma vez que o array inteiro é copiado. No caso de uma slice, não há um custo real
+uma vez que somente os valores internos da slice são copiados e não o array de suporte. Segundo, você recebe uma copia de 
+todos os elementos sendo iterados.
```
Comment (rschio) [inline _content/tour/por/arrays.article:68]:
```suggestion
Quando iteramos usando valor, duas coisas acontecem. Primeira, a collection inteira é copiada e você navega sobre a cópia.
No caso de arrays, a cópia pode ser bem cara uma vez que o array inteiro é copiado. No caso de uma slice, não há um custo real
uma vez que somente os valores internos da slice são copiados e não o array de suporte. Segundo, você recebe uma cópia de 
todos os elementos sendo iterados.
```

File: _content/tour/por/arrays.article [hunk 3]
```diff
@@ -1,163 +1,153 @@
...
     4 Plum
 
-The strings variable is an array of 5 strings. The loop iterates over each string
-in the collection and displays the index position and the string value. Since this
-is value semantic iteration, the for range is iterating over its own shallow copy
-of the array and on each iteration the fruit variable is a copy of each string
-(the two word data structure).
+A variável strings é um array de tamanho 5 do tipo string. O loop passa por cada string da collection
+e mostra o índice da posição e o valor da string. Como essa iteração é por valor, o for range itera por uma shallow copy 
+do array e cada iteração a variável fruit é uma copia de cada string (a estrutura de dados de duas palavras).
```
Comment (rschio) [inline _content/tour/por/arrays.article:91]:
```suggestion
do array e cada iteração a variável fruit é uma cópia de cada string (a estrutura de dados de duas palavras).
```

File: _content/tour/por/arrays.article [hunk 4]
```diff
@@ -1,163 +1,153 @@
...
+
+Perceba como a variável fruit é passada para a função print usando valores. A função print também está pegando 
+sua própria copia do valor da string. Quando a string é passada para a função print, temos 4 copias do valor da 
+string (array, shallow copy, variável fruit e copia para a função print). Todas as 4 copias estão compartilhando 
+o mesmo array de bytes.
```
Comment (rschio) [inline _content/tour/por/arrays.article:96]:
```suggestion
Perceba como a variável fruit é passada para a função print usando valores. A função print também está pegando 
sua própria cópia do valor da string. Quando a string é passada para a função print, temos 4 cópias do valor da 
string (array, shallow copy, variável fruit e a cópia para a função print). Todas as 4 cópias estão compartilhando 
o mesmo array de bytes.
```

File: _content/tour/por/arrays.article [hunk 5]
```diff
@@ -1,163 +1,153 @@
...
-The print function is getting its own copy of the string value as well. By the time
-the string is passed to the print function, there are 4 copies of the string value
-(array, shallow copy, fruit variable and the print function’s copy). All 4 copies
-are sharing the same backing array of bytes. 
 
 .image /tour/eng/static/img/a3.png
 
-Making copies of the string value is important because it prevents the string value
-from ever escaping to the heap. This eliminates non-productive allocation on the heap.
+Fazer copias do valor da string é importante para prevenir que o valor da string acabe vazando para o heap.
```
Comment (rschio) [inline _content/tour/por/arrays.article:101]:
```suggestion
Fazer cópias do valor da string é importante para prevenir que o valor da string acabe vazando para o heap. 
```

File: _content/tour/por/arrays.article [hunk 6]
```diff
@@ -179,79 +169,84 @@ semantic iteration to display information about each string. The output shows
...
+
+Note como o array é um bloco de memória contígua e como a string tem 2 palavras ou 
+uma estrutura de dados de 16 bytes dentro da minha arquitetura de 64 bits.
+O endereço de cada elemento acontece em intervalos de 16 bytes.
 
-The fact that the v variable has the same address on each iteration strengthens the
-understanding that v is a local variable of type string which contains a copy of each
-string value during iteration.
+O fato de que a variável v tenha o mesmo endereço em cada iteração, fortalece o 
+entendimento de que v é uma variável local de tipo string que contém a copia de
```
Comment (rschio) [inline _content/tour/por/arrays.article:182]:
```suggestion
entendimento de que v é uma variável local de tipo string que contém a cópia de 
```

File: _content/tour/por/arrays.article [hunk 7]
```diff
@@ -179,79 +169,84 @@ semantic iteration to display information about each string. The output shows
...
+cada valor da string durante a iteração.
 
-** CPU Caches
+** Cache de CPU
 
-There are lots of mechanical differences between processors and their design. In
-this section, you will talk at a high level about processors and the semantics that
-are relatively the same between them all. This semantic understanding will provide
-you a good mental model for how the processor works and the sympathy you can provide.
+Existem várias mecânicas diferentes entre processadores e seus designes. Nesta seção,
```
Comment (rschio) [inline _content/tour/por/arrays.article:187]:
```suggestion
Existem várias mecânicas diferentes entre processadores e seus designs. Nesta seção, 
```

File: _content/tour/por/arrays.article [hunk 8]
```diff
@@ -179,79 +169,84 @@ semantic iteration to display information about each string. The output shows
...
 
-** CPU Caches
+** Cache de CPU
 
-There are lots of mechanical differences between processors and their design. In
-this section, you will talk at a high level about processors and the semantics that
-are relatively the same between them all. This semantic understanding will provide
-you a good mental model for how the processor works and the sympathy you can provide.
+Existem várias mecânicas diferentes entre processadores e seus designes. Nesta seção, 
+falaremos, em alto nível, sobre processadores e as semânticas que sao relativamente
```
Comment (rschio) [inline _content/tour/por/arrays.article:188]:
```suggestion
falaremos, em alto nível, sobre processadores e as semânticas que são relativamente
```

File: _content/tour/por/arrays.article [hunk 9]
```diff
@@ -179,79 +169,84 @@ semantic iteration to display information about each string. The output shows
...
+as mesmas entre todos eles. O entendimento dessa semântica te dará um bom modelo 
+mental de como o processador funciona e a empatia que você pode oferecer.
 
-Each core inside the processor has its own local cache of memory (L1 and L2) and a 
-common cache of memory (L3) used to store/access data and instructions. The hardware
-threads in each core can access their local L1 and L2 caches. Data from L3 or main
-memory needs to be copied into the L1 or L2 cache for access.
+Cada core do processador tem sua própria memoria cache (L1 e L2) e uma memória cache
```
Comment (rschio) [inline _content/tour/por/arrays.article:192]:
```suggestion
Cada core do processador tem sua própria memória cache (L1 e L2) e uma memória cache 
```

File: _content/tour/por/arrays.article [hunk 10]
```diff
@@ -297,82 +294,84 @@ for cache line prefetching and becomes essentially random access memory.
...
+Cada programa em execução recebe um mapa completo da memória virtual do Sistema operacional
+e o programa pensa que ele tem toda a memória física da máquina. Porém, a memória física 
+precisa ser compartilhada com todos os programas em execução. O sistema operacional 
+compartilha a memória física dividindo-a em páginas e alocando essas páginas à programas 
+em execução. Cada sistema operacional decide o tamanho dessas páginas,  mas 4k, 8k e 16k 
+são tamanhos razoáveis e comumente utilizados.   
+
+O TLB é um pequeno cache dentro do processador que ajuda a reduzir a latência da tradução do 
+endereço da memória virtual para o endereço da memória física dentro do escopo de uma página 
+do sistema operacional e de seu descolamento dentro da página. Uma falha no cache TLB pode
```
Comment (rschio) [inline _content/tour/por/arrays.article:317]: "Descolamento" is weird, I think "offset" fits better.

File: _content/tour/por/arrays.article [hunk 11]
```diff
@@ -297,82 +294,84 @@ for cache line prefetching and becomes essentially random access memory.
...
-    Round trip within same datacenter ...... 500,000 ns (0.5 ms) .......... 6M ins
-    Read 1 MB sequentially from SSD- ..... 1,000,000 ns (1 ms) ........... 12M ins
-    Disk seek ........................... 10,000,000 ns (10 ms) ......... 120M ins
-    Read 1 MB sequentially from disk .... 20,000,000 ns (20 ms) ......... 240M ins
-    Send packet CA->Netherlands->CA .... 150,000,000 ns (150 ms) ........ 1.8B ins
+    Referencia de cache L1 ........................ 0.5 ns ...................  6 ins
+    Branch mispredict ............................... 5 ns ................... 60 ins
+    Referencia de cache L2  ......................... 7 ns ................... 84 ins
+    Mutex lock/unlock .............................. 25 ns .................. 300 ins
+    Referencia da memória principal ............... 100 ns ................. 1200 ins
```
Comment (rschio) [inline _content/tour/por/arrays.article:357]:
```suggestion
    Referência de cache L1 ........................ 0.5 ns ...................  6 ins
    Branch mispredict ............................... 5 ns ................... 60 ins
    Referência de cache L2  ......................... 7 ns ................... 84 ins
    Mutex lock/unlock .............................. 25 ns .................. 300 ins
    Referência da memória principal ............... 100 ns ................. 1200 ins           
```

File: _content/tour/por/arrays.article [hunk 12]
```diff
@@ -402,27 +401,26 @@ how efficiently you can get data into the processor.
...
-** Exercise 1
+** Exercício 1
 
-Declare an array of 5 strings with each element initialized to its zero value. Declare
-a second array of 5 strings and initialize this array with literal string values. Assign
-the second array to the first and display the results of the first array. Display the
-string value and address of each element.
+Declare um array com 5 strings com cada elemento inicializado com o seu valor inicial padrão.
+Declare um segundo array com 5 strings e o inicializa com valores literais para as strings. 
+Atribua o segundo array ao primeiro e mostre o resultado of primeiro array. Imprima o valor da string
```
Comment (rschio) [inline _content/tour/por/arrays.article:422]:
```suggestion
Declare um segundo array com 5 strings e o inicialize com valores literais para as strings. 
Atribua o segundo array ao primeiro e mostre o resultado do primeiro array. Imprima o valor da string
```

File: _content/tour/por/arrays.article [hunk 13]
```diff
@@ -179,79 +169,84 @@ semantic iteration to display information about each string. The output shows
 each individual string value, the address of the v variable and the address of
```
Comment (rschio) [inline _content/tour/por/arrays.article:169]: Here, you forgot to delete the english text.

File: _content/tour/por/arrays.article [hunk 14]
```diff
@@ -179,79 +169,84 @@ semantic iteration to display information about each string. The output shows
...
+** Cache de CPU
 
-There are lots of mechanical differences between processors and their design. In
-this section, you will talk at a high level about processors and the semantics that
-are relatively the same between them all. This semantic understanding will provide
-you a good mental model for how the processor works and the sympathy you can provide.
+Existem várias mecânicas diferentes entre processadores e seus designes. Nesta seção, 
+falaremos, em alto nível, sobre processadores e as semânticas que sao relativamente
+as mesmas entre todos eles. O entendimento dessa semântica te dará um bom modelo 
+mental de como o processador funciona e a empatia que você pode oferecer.
```
Comment (rschio) [inline _content/tour/por/arrays.article:190]: I think here should be simpatia not empatia.

File: _content/tour/por/arrays.article [hunk 15]
```diff
@@ -179,79 +169,84 @@ semantic iteration to display information about each string. The output shows
...
+Go usam um array por trás.
+
+Quando você constrói um array, todo elemento está igualmente distante do seu elementos anterior 
+e seguinte. A medida que você itera um array, você começa a navegar de uma linha de cache para 
+uma linha de cache conectada de forma previsível. O Prefetcher vai então se utilizar desse padrão 
+de acesso previsível e irá, de forma eficiente, trazer os dados para o processador, reduzindo 
+assim o custo de latência. 
+
+Imagine agora que você tem uma matriz quadrática de memória e uma lista encadeada com a mesma 
+quantidade de nós que a matrix. Se você executar uma travessia na lista encadeada e na matriz
```
Comment (rschio) [inline _content/tour/por/arrays.article:248]:
```suggestion
quantidade de nós que a matriz. Se você executar uma travessia na lista encadeada e na matriz 
```

**Review by rschio — COMMENTED** [review]
There is a file "array_list.html" which contains the part from "Notas sobre o cache de CPU", maybe we should translate.

**Comment by sergiolpf** [comment]
> There is a file "array_list.html" which contains the part from "Notas sobre o cache de CPU", maybe we should translate.

Done!

**Review by rschio — APPROVED** [review]

**Pull request merged**